[WebGPU] index buffer support

mwyrzykowski opened the pull request:
#### 2eaacb29a834ef362921d003b78d73e1354fbcf7
<pre>
[WebGPU] index buffer support
<a href="https://bugs.webkit.org/show_bug.cgi?id=249238">https://bugs.webkit.org/show_bug.cgi?id=249238</a>
&lt;radar://103240396&gt;

Reviewed by Myles C. Maxfield and Dean Jackson.

Implement index buffer support and add a test case.

Also implement setScissorRect. Everything was implemented
except the call in WebGPU.

* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::drawIndexed):
(WebGPU::RenderPassEncoder::setIndexBuffer):
Call respective metal commands for drawing indexed primitives.

(WebGPU::RenderPassEncoder::setScissorRect):
Implement setScissorRect with coordinates specified in pixels.
* Websites/webkit.org/demos/webgpu/hello-indexed-cube.html: Added.
* Websites/webkit.org/demos/webgpu/scripts/hello-indexed-cube.js: Added.
(async helloCube.frameUpdate):
(async helloCube):
Rotating cube with 8 vertices and 36 indices.

Canonical link: <a href="https://commits.webkit.org/258168@main">https://commits.webkit.org/258168@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ca4fec9b387421f7918447669fd365da072da33

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101060 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10216 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34114 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110361 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11145 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1092 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93472 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108194 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106843 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8440 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/35053 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23103 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3882 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24620 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3911 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/1023 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10022 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44128 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5603 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5678 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WebGPU/RenderPassEncoder.mm:
```diff
@@ -201,10 +198,7 @@
-    UNUSED_PARAM(y);
-    UNUSED_PARAM(width);
-    UNUSED_PARAM(height);
+    [m_renderCommandEncoder setScissorRect: { x, y, width, height } ];
```
IIRC there are corner cases that need to be checked (like if width is negative or if x > viewport or something like that). Maybe we should add a FIXME.

litherum reviewed (approved):
Any progress on the CTS with this change?

mwyrzykowski committed:
[WebGPU] index buffer support
https://bugs.webkit.org/show_bug.cgi?id=249238
<radar://103240396>

Reviewed by Myles C. Maxfield and Dean Jackson.

Implement index buffer support and add a test case.

Also implement setScissorRect. Everything was implemented
except the call in WebGPU.

* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::drawIndexed):
(WebGPU::RenderPassEncoder::setIndexBuffer):
Call respective metal commands for drawing indexed primitives.

(WebGPU::RenderPassEncoder::setScissorRect):
Implement setScissorRect with coordinates specified in pixels.
* Websites/webkit.org/demos/webgpu/hello-indexed-cube.html: Added.
* Websites/webkit.org/demos/webgpu/scripts/hello-indexed-cube.js: Added.
(async helloCube.frameUpdate):
(async helloCube):
Rotating cube with 8 vertices and 36 indices.

Canonical link: https://commits.webkit.org/258168@main
